File: src/desphere/wav.py
"""Minimal canonical RIFF/WAVE (PCM) writer.

Pure stdlib ``struct``; emits the standard 44-byte header (RIFF / ``fmt `` /
``data``) with little-endian integer PCM samples. The WAV format is an open,
publicly documented container (Microsoft/IBM RIFF spec); no GPL source involved.
"""

from __future__ import annotations

import struct

from .errors import DesphereError

PCM_FORMAT = 1  # WAVE_FORMAT_PCM


def write_wav(
    out,
    *,
    channels: int,
    sample_rate: int,
    bits_per_sample: int,
    data: bytes,
) -> None:
    """Write a canonical PCM WAV to the binary stream ``out``.

    ``data`` must already be little-endian, interleaved PCM at the given
    ``bits_per_sample``.
    """
    if bits_per_sample % 8 != 0:
        raise ValueError(f"bits_per_sample must be a multiple of 8, got {bits_per_sample}")

    bytes_per_sample = bits_per_sample // 8
    block_align = channels * bytes_per_sample
    byte_rate = sample_rate * block_align
    data_size = len(data)
    pad = data_size & 1  # data chunk must be word-aligned

    riff_size = 4 + (8 + 16) + (8 + data_size + pad)

    # RIFF/WAV stores sizes in unsigned 32-bit fields. Fail loud BEFORE writing
    # any bytes (so we never leave a truncated "RIFF" stub) rather than letting
    # struct.pack raise an opaque struct.error mid-write. riff_size >= data_size,
    # so this one check covers both 32-bit size fields.
    if riff_size > 0xFFFFFFFF:
        raise DesphereError(
            "output exceeds the 4 GB RIFF/WAV size limit "
            f"({data_size} bytes of PCM overflow the 32-bit size fields)"
        )

    out.write(b"RIFF")
    out.write(struct.pack("<I", riff_size))
    out.write(b"WAVE")

    out.write(b"fmt ")
    out.write(
        struct.pack(
            "<IHHIIHH",
            16,               # fmt chunk size
            PCM_FORMAT,       # audio format
            channels,
            sample_rate,
            byte_rate,
            block_align,
            bits_per_sample,
        )
    )

    out.write(b"data")
    out.write(struct.pack("<I", data_size))
    out.write(data)
    if pad:
        out.write(b"\x00")

File: rust/src/wav.rs
//! Minimal canonical RIFF/WAVE (PCM) writer — Rust twin of `src/desphere/wav.py`.
//! Emits the standard 44-byte header (RIFF / `fmt ` / `data`) + little-endian PCM.

use crate::error::DecodeError;

const PCM_FORMAT: u16 = 1;

/// Build a canonical PCM WAV. `data` must already be little-endian interleaved
/// PCM at `bits_per_sample`.
pub fn write_wav(
    channels: u16,
    sample_rate: u32,
    bits_per_sample: u16,
    data: &[u8],
) -> Result<Vec<u8>, DecodeError> {
    if bits_per_sample & 7 != 0 {
        return Err(DecodeError::Corrupt(format!(
            "bits_per_sample must be a multiple of 8, got {bits_per_sample}"
        )));
    }
    let bytes_per_sample = (bits_per_sample / 8) as u32;
    let block_align = channels as u32 * bytes_per_sample; // <= 65535*4, fits u32
                                                          // u64 math so the size checks are real on 32-bit targets (wasm32: usize=u32).
    let byte_rate = sample_rate as u64 * block_align as u64;
    if byte_rate > u32::MAX as u64 {
        return Err(DecodeError::Corrupt(
            "WAV byte_rate exceeds 32-bit field".into(),
        ));
    }
    let data_size = data.len();
    let pad = data_size & 1; // data chunk must be word-aligned

    let riff_size = 4u64 + (8 + 16) + (8 + data_size as u64 + pad as u64);
    if riff_size > 0xFFFF_FFFF {
        return Err(DecodeError::Corrupt(format!(
            "output exceeds the 4 GB RIFF/WAV size limit ({data_size} bytes of PCM)"
        )));
    }

    let mut out = Vec::with_capacity(44 + data_size + pad);
    out.extend_from_slice(b"RIFF");
    out.extend_from_slice(&(riff_size as u32).to_le_bytes());
    out.extend_from_slice(b"WAVE");
    out.extend_from_slice(b"fmt ");
    out.extend_from_slice(&16u32.to_le_bytes()); // fmt chunk size
    out.extend_from_slice(&PCM_FORMAT.to_le_bytes());
    out.extend_from_slice(&channels.to_le_bytes());
    out.extend_from_slice(&sample_rate.to_le_bytes());
    out.extend_from_slice(&(byte_rate as u32).to_le_bytes());
    out.extend_from_slice(&(block_align as u16).to_le_bytes());
    out.extend_from_slice(&bits_per_sample.to_le_bytes());
    out.extend_from_slice(b"data");
    out.extend_from_slice(&(data_size as u32).to_le_bytes());
    out.extend_from_slice(data);
    if pad == 1 {
        out.push(0);
    }
    Ok(out)
}
